src/Tracing.jl

```diff
@@ -201,9 +125,88 @@ function traced_type(::Type{T}, seen, mode, track_numbers) where {T}
     throw(NoFieldMatchError(T, TT2))
 end
 
-function traced_type(
-    ::Type{<:ConcreteRNumber{T}}, seen, ::Val{mode}, track_numbers
-) where {T,mode}
+for T in (DataType, Module, Nothing, Symbol, AbstractChar, AbstractString, AbstractFloat, Integer, RNumber)
+    @eval Base.@nospecializeinfer function traced_type(@nospecialize(T::Type{<:$T}), seen, mode, track_numbers)
+        return T
+    end
+end
+
+Base.@nospecializeinfer function traced_type(
+    @nospecialize(T::Type{<:ReactantPrimitive}), seen, mode::Val{Mode}, track_numbers
+)
+    if Mode == ArrayToConcrete && any(Base.Fix1(<:, T), track_numbers)
+        return ConcreteRNumber{T}
+    end
+    return T
+end
+
+Base.@nospecializeinfer function traced_type(
+    @nospecialize(C::Type{<:Complex}), seen::ST, mode::Val{Mode}, track_numbers::TN
+) where {ST,Mode,TN}
+    if !(C isa UnionAll)
+        return Complex{traced_type(C.parameters[1], seen, mode, track_numbers)}
+    else
+        return C
+    end
+end
+
+Base.@nospecializeinfer function traced_type(@nospecialize(T::Type{<:Function}), seen, mode, track_numbers)
+    # functions are directly returned
+    if sizeof(T) == 0
+        return T
+    end
+
+    # in closures, enclosured variables need to be traced
+    N = fieldcount(T)
+    changed = false
+    traced_fieldtypes = ntuple(Val(N)) do i
+        next = traced_type(fieldtype(T, i), seen, mode, track_numbers)
+        changed |= next != fieldtype(T, i)
+        next
+    end
+
+    if !changed
+        return T
+    end
+
+    # closure are struct types with the types of enclosured vars as type parameters
+    return Core.apply_type(T.name.wrapper, traced_fieldtypes...)
+end
+
+@inline is_concrete_tuple(x::T2) where {T2} =
+    (x <: Tuple) && !(x === Tuple) && !(x isa UnionAll)
```

Comment:
isn't `T2` just `Type`? and because `Type` is parametric, we could do `@nospecialize` and maybe `@nospecializeinfer` too (it always returns `Bool`)? wdyt @gbaraldi?

```suggestion
@inline @nospecializeinfer is_concrete_tuple(@nospecialize(T)) =
    (T <: Tuple) && !(T === Tuple) && !(T isa UnionAll)
```